### PR TITLE
Enable user to provide ClientWebSocket instance

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**
## Checklist
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/main/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [ ] This pull-request is submitted against the `main` branch.

### CODE REVIEW

File: iothub/device/src/Transport/Amqp/ClientWebSocketTransport.cs [hunk 1]
```diff
@@ -358,7 +358,6 @@ private static void ValidateBufferBounds(byte[] buffer, int offset, int size)
             }
         }
 
-
```
Comment (abhipsaMisra) [inline iothub/device/src/Transport/Amqp/ClientWebSocketTransport.cs:361]: I also recommend updating the namespace this class is under from an Azure.Amqp namespace to one under Microsoft.Azure.Devices.Client.

File: iothub/device/src/Transport/AmqpIot/AmqpIotTransport.cs [hunk 2]
```diff
@@ -124,38 +122,13 @@ private async Task<ClientWebSocket> CreateClientWebSocketAsync(Uri websocketUri,
                 if (Logging.IsEnabled)
                     Logging.Enter(this, nameof(CreateClientWebSocketAsync));
 
-                var websocket = new ClientWebSocket();
+                bool customWebSocket = _amqpTransportSettings.ClientWebSocket != null;
+                ClientWebSocket websocket = customWebSocket ? _amqpTransportSettings.ClientWebSocket : new ClientWebSocket();
+                IWebProxy webProxy = customWebSocket ? websocket.Options.Proxy : _amqpTransportSettings.Proxy;
```
Comment (abhipsaMisra) [inline iothub/device/src/Transport/AmqpIot/AmqpIotTransport.cs:127]: Where is this webProxy being used?

File: iothub/device/src/Transport/AmqpIot/AmqpIotTransport.cs [hunk 3]
```diff
@@ -124,38 +122,13 @@ private async Task<ClientWebSocket> CreateClientWebSocketAsync(Uri websocketUri,
...
-                    }
-                }
-                catch (PlatformNotSupportedException)
-                {
-                    // .NET Core 2.0 doesn't support proxy. Ignore this setting.
-                    if (Logging.IsEnabled)
-                        Logging.Error(this, $"{nameof(CreateClientWebSocketAsync)} PlatformNotSupportedException thrown as .NET Core 2.0 doesn't support proxy");
-                }
-
-                if (_amqpTransportSettings.WebSocketKeepAlive.HasValue)
```
Comment (abhipsaMisra) [inline iothub/device/src/Transport/AmqpIot/AmqpIotTransport.cs:152]: Is this no longer being set on the webSocket? The customer likely won't know to set this, so we should check their supplied websocket instance to see if they have set it, else set it for them. In the case where we create the web socket for them, we still need to set it.

File: iothub/device/src/TransportSettings/IotHubClientAmqpSettings.cs [hunk 4]
```diff
@@ -72,25 +73,19 @@ public TimeSpan OperationTimeout
         }
 
         /// <summary>
-        /// A keep-alive for the transport layer in sending ping/pong control frames when using web sockets.
-        /// </summary>
-        /// <seealso href="https://docs.microsoft.com/dotnet/api/system.net.websockets.clientwebsocketoptions.keepaliveinterval"/>
-        public TimeSpan? WebSocketKeepAlive { get; set; }
```
Comment (abhipsaMisra) [inline iothub/device/src/TransportSettings/IotHubClientAmqpSettings.cs:78]: Removed?

File: iothub/device/src/Transport/AmqpIot/AmqpIotTransport.cs [hunk 5]
```diff
@@ -49,8 +48,7 @@ internal class AmqpIotTransport : IDisposable
             {
                 TargetHost = hostName,
                 Certificate = null,
-                CertificateValidationCallback = _amqpTransportSettings.RemoteCertificateValidationCallback
-                    ?? OnRemoteCertificateValidation,
+                CertificateValidationCallback = OnRemoteCertificateValidation,
```
Comment (abhipsaMisra) [inline iothub/device/src/Transport/AmqpIot/AmqpIotTransport.cs:51]: Ah, this looks like a gap in our plan. The remote certificate validation callback is being set into the TLS settings, so we'll need a way for users to be able to pass this to us for tcp-based connections as well.

File: iothub/device/src/TransportSettings/IotHubClientAmqpSettings.cs (hunk 4 above)
Comment (drwill-ms) [reply iothub/device/src/TransportSettings/IotHubClientAmqpSettings.cs:78]: Are we taking the path that if they want this stuff set, they provide the whole instance?

Comment (drwill-ms) [reply iothub/device/src/TransportSettings/IotHubClientAmqpSettings.cs:78]: And is that consistent with our similar situations, e.g., HttpClient?

File: iothub/device/src/Edge/CustomCertificateValidator.cs [hunk 6]
```diff
@@ -41,15 +41,7 @@ private void SetupCertificateValidation()
...
-            if (_transportSettings is IotHubClientAmqpSettings amqpTransportSettings)
-            {
-                if (amqpTransportSettings.RemoteCertificateValidationCallback == null)
-                {
-                    amqpTransportSettings.RemoteCertificateValidationCallback =
-                        (sender, certificate, chain, sslPolicyErrors) => ValidateCertificate(_certs.First(), certificate, chain, sslPolicyErrors);
-                }
-            }
-            else if (_transportSettings is IotHubClientMqttSettings mqttTransportSettings)
+            if (_transportSettings is IotHubClientMqttSettings mqttTransportSettings)
```
Comment (drwill-ms) [inline iothub/device/src/Edge/CustomCertificateValidator.cs:44]: Why is AMQP path being removed and not MQTT?

**Review by drwill-ms — CHANGES_REQUESTED** [review]
ClientWebSocket inherits from WebSocket which is IDisposable. What is our plan for disposing this when our client is disposed? Do we always dispose it? Do we leave that to the user to do? Do we tell them that? Do we give them an option, like other .NET cases do?

https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket?view=net-6.0

**Pull request closed without merging**